### DISCUSSION TITLE
Add mod where every hit object appears at once

### DISCUSSION THREAD

**Discussion by mk56-spn** [ppy/osu#16732] [Ideas]
https://www.youtube.com/watch?v=32T3shB6VTs&t=21s&ab_channel=IsakVik

a niche but established playstyle that has a history,  would be nice to see it added, dunno where it would go since approach circles dont make sense here but if its under DA (at the beginning of the AR bar) it would be weird for a DA setting to disable approach circles, anyways im busy RN but like id be curious to see what people make of this

**Comment by peppy** [comment]
The video you linked shows something. But that something is not "negative infinity approach rate".

It looks like a new mod idea named something appropriate, but is not related to approach rate in any way.

**Reply by mk56-spn** [reply]
its the common name given to this by the community. whether technically correct or not

**Reply by peppy** [reply]
And won't be a term we use, because it makes negative infinite sense.

**Reply by mk56-spn** [reply]
Good one. I rest my case..any ideas?..something like "clusterfuck" but less rude might work idk

**Reply by hlysine** [reply]
What about "Piled Up"?

**Reply by mk56-spn** [reply]
> What about "Piled Up"?

sounds palatable.

**Comment by NotGumballer91** [comment]
You mean like every circle would appear and every circle would have different AR?

**Reply by mk56-spn** [reply]
approach circles would be hidden but yes i guess

**Comment by mk56-spn** [comment]


**Reply by mk56-spn** [reply]
![osu_2022-02-01_17-52-50](https://user-images.githubusercontent.com/74463310/152018637-d7fccfd1-003c-42e2-b4fd-0565264c556f.png)

**Reply by bdach** [reply]
i'm not honestly sure what you were expecting, or why are you pinging random contributors to help you

about the only thing that could save you is forcing the sliders to be opaque to avoid having to draw new sliders over old ones and even that isn't guaranteed to help all that much because of transparency and shadow effects around the path

or just limiting the number of objects to next $number only i guess

**Reply by peppy** [reply]
The issue is not “laser’s render pipeline”. The issue is your lacking implementation. As is already said, you can’t throw shit and expect it to work.

A mod like this requires due diligence to implement.

**Reply by smoogipoo** [reply]
Yes, as you've found out, this is likely one of the _most_ pathological cases you can get for osu!lazer/osu!framework, where everything will begin to fall apart and all the little costs from things like pooling, updating children life outside of a LifetimeEntryManager, scheduler updates/invocations, and running multiple catchup update frames during replay playback, will all add up.

Even just our vertex batching heuristics are simply not good enough for this, and is part of the reason why multiplayer spectator is limited to 16 instances.

**Reply by peppy** [reply]
For anyone looking to actually implement something like this, a better way forward would be drawing the non-active objects to some kind of buffer, rather than attempting to make every hitobject alive, handling input and drawing/updating every frame.

**Reply by hlysine** [reply]
> For anyone looking to actually implement something like this, a better way forward would be drawing the non-active objects to some kind of buffer, rather than attempting to make every hitobject alive, handling input and drawing/updating every frame.

I implemented this idea in https://github.com/hlysine/osu/tree/osu-piled-up-mod and managed to get the game running smoothly while displaying all hit objects (most are displayed via frame buffers).

![image](https://user-images.githubusercontent.com/25472513/152713443-e8bb5cfd-f664-4772-8647-35a10a819a2f.png)

However, I highly doubt that this would ever be worth a PR, since there are too many issues and dirty hacks involved:
- major stuttering at the start of the map when the framebuffers are being prepared (see screenshot above)
- a dirty hack to clear children of `BufferedContainer`s after they've been rendered to buffer (creating a custom buffered container is probably the better way)
- minor stutters and flickers during gameplay when a buffer is being "converted" to real hit objects
- buffered circles not animated (Kiai time flash, slider repeat circle pulse, etc)
- buffered circles not shaking when clicked (not a dealbreaker imo)
- skin change during gameplay not supported
- rewinding not supported (should totally be possible to support, just don't know how) 

I'm not spending any more time on this for now. Just documenting my results for anyone interested to pick up. A note is included in the mod class for detailed explanation of what's going on.

**Reply by smoogipoo** [reply]
@hlysine you have the correct idea.

Stutters on load could be made better by having a higher level component that mods can queue things to be drawn during the load sequence of `Player`. Have it wait draw 1 new drawable per frame and influence the load time of `Player`. Implementation would be similar to screenshotting: https://github.com/ppy/osu-framework/blob/4de8d3c1ad3f73d4b9fb0f8902127eaf010d406f/osu.Framework/Platform/GameHost.cs#L543-L576.

**Reply by hlysine** [reply]
Thanks, that's exactly what I was looking for. In that case, I'll spend a bit more time on this implementation and see if I can get it to a PR-ready state.

**Comment by mk56-spn** [comment]
alright so i wanted to try a serious implementation of this with a bit more knowledge now. but id appreciate if i can have some questions answered regarding how feasible a few things are:

- for the loading of the real hit objects and the discarding of their fake placeholder versions when they are close to needing to be hit would it be feasible and ( and does the framework even support i should ask) reserving a thread for all this, kind of how im under the impression the inputs and audio are separated into threads to lower latency

- a dynamic buffer that looks for gaps in game play within a certain range of where the next "bufferedContainer" should be loaded in and delays / expedites the rendering of the objects as to make micro stutters, if totally unavoidable, cause the least amount of pain to the player

- this one is drastic but the only thing i could think of for low end systems, flatten all rendered objects into successive pngs covering the whole visible playfield at each buffer point, for the fake hit objects to create the mods effect, and hide and load these progressively as the map goes on showing each next one with the same objects  - ( the ones that have already been hit + the ones being actively rendered). obviously this wouldnt allow for any reactivity from the notes but idk its the best ive been able to think of to reduce strain on the system, this ofcourse supposes theres even a way to snapshot the playfield, which im assuming can theoretically be done since lazer has a screenshot system

**Reply by peppy** [reply]
I'm not really willing to provide any direction forward for this one. Try things out and see what works. It will need experimentation before choosing a path forward.

I will say that loading PNGs is so slow it will likely be worse performance than anything else.

**Reply by astro-angelfish** [reply]
But I'd like to say the last idea does not seem to be easy to do because taking screenshot requires another game window. osu!framework didn't prepare anything for headless rendering.

I do still remember I have did something related to headless rendering by similar framework called lwjgl with EGL in java on my own server side project. Here is my (shit-like?) code: https://gist.github.com/astro-angelfish/45cf9bfbd6cc1f838f328fe304cd19b2

Maybe we could do the same thing on osu!framework to enable headless rendering.

**Reply by peppy** [reply]
> But I'd like to say the last idea does not seem to be easy to do because taking screenshot requires another game window

Except screenshots can be taken of any backbuffer (already implemented) and it's not a huge reach to do the same at a local level. Please check things out before making statements like this, especially when bumping old threads. It just makes noise.
